src/examples/learn_quant/monotonicity.py

```diff
@@ -0,0 +1,254 @@
+from ultk.effcomm.informativity import informativity
+from ultk.language.grammar import GrammaticalExpression
+from ultk.language.language import Language, aggregate_expression_complexity
+from ultk.language.semantics import Meaning
+
+from learn_quant.meaning import create_universe
+from learn_quant.util import read_expressions
+from learn_quant.quantifier import QuantifierUniverse, QuantifierModel
+from concepts.contexts import Context
+
+from itertools import product
+import math
+from scipy.stats import entropy
+import pandas as pd
+import numpy as np
+from tqdm import tqdm
+from scipy import sparse
+
+from pathlib import Path
+
+import hydra
+from omegaconf import DictConfig, OmegaConf
+
+
+def switch_direction(bools: np.ndarray, down=False) -> np.ndarray:
+    """Switches the value of True to False and vice versa.
+
+    Args:
+        bool (np.ndarray): Accepts a list of lists of bools
+
+    Returns:
+        np.ndarray: A list of lists of bools, flipped unless down == True.
+    """
+    if down:
+        return bools
+    else:
+        return bools
+
+
+def get_truth_matrix(universe: QuantifierUniverse) -> np.ndarray:
+    """Get a numpy matrix that stores the truth vectors of all models in the quantifier universe
+
+    Returns:
+        np.ndarray: A matrix of booleans arrays that track the indices of set items of A intersected by B.
+    """
+    truth_array = []
+    names_array = []
+    for quantifier_model in universe.referents:
+        truth_vector = tuple(True if x == "2" else False for x in quantifier_model.name)
+        truth_array.append(truth_vector)
+        names_array.append(quantifier_model.name)
+    truth_matrix = np.array(truth_array)
+    names_vector = np.array(names_array)
+    return truth_matrix, names_vector
+
+
+def calculate_lattice(universe: QuantifierUniverse, down=False) -> Context:
+    """
+
+    Returns:
+        Context: Created from ``objects``, ``properties``, and ``bools`` correspondence.
+
+        objects: Iterable of object label strings.
+        properties: Iterable of property label strings.
+        bools: Iterable of ``len(objects)`` tuples of ``len(properties)`` booleans.
+
+        See https://github.com/haberchr/ttps://github.com/haberchr/concepts/blob/master/concepts/contexts.py
+
+    """
+    truth_df = get_truth_matrix(universe)
+    objects = names_vector.tolist()
+    properties = truth_df.columns
+    bools = switch_direction(truth_df, down)
+    return Context(objects, properties, bools)
+
+
+def get_sub_structures(concept_lattice: Context, name: list[str]) -> set[str]:
```

Comment:
Let's walk through the logic of the monotonicity calculation in our next meeting and discuss the `-set(name)` part.  In general, the substructure relation in our definition is _reflexive_, i.e. every structure is a sub-structure of itself; but thtere might be reasons in the way you have things setup to exclude that